### PR TITLE
True half-precision support in Trainer

### LINKED ISSUE #17609: Support true half-precision in Trainer
### Description & Motivation

We recently added the `HalfPrecision` plugin in Fabric. The usage looks like this:

```py
fabric = Fabric(precision="16-true")
# or
fabric = Fabric(precision="bf16-true")
```

This feature is not exclusive to Fabric and can easily be added to Trainer as well.

### Pitch

Support the following:

```py
trainer = Trainer(precision="16-true")
# or
trainer = Trainer(precision="bf16-true")
```

To support this, we need to add the HalfPrecision plugin and minor modifications to the DeepSpeedPrecision and FSDPPrecision plugins. 

### Alternatives

_No response_

### Additional context

This change should also include/close #17727

cc @borda @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Implements true half-precision support in Trainer, with corresponding changes as done for Fabric in #17287.

Fixes #17609 

Follow up work:
- #17727, #17728 
- FSDP
- DeepSpeed
- XLA #18172

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @justusschock @awaelchli

### CODE REVIEW

File: docs/source-pytorch/common/trainer.rst
Comment (justusschock) [inline docs/source-pytorch/common/trainer.rst:813]:
```suggestion
Half precision is using 16 bit floating point operations while mixed precision is the combined use of 32 and 16 bit floating points to reduce memory footprint during model training. Since not all operations (like batchnorm) are numerically stable in lower bit precisions, these operations will still be carried out in fp32 whereas half precision unconditionally performs all operations in 16 bit.
```

**Review by justusschock — APPROVED** [review]
minor comment to clearly differentiate between half and mixed precision. Wording wasn't clear to me before

**Review by Borda — APPROVED** [review]

**Review by lantiga — APPROVED** [review]
LGTM, after merging @justusschock 's suggestion on the description

**Pull request merged**
